pkg/controller/keystoneservice/keystoneservice_controller.go

```diff
@@ -0,0 +1,163 @@
+var _ reconcile.Reconciler = &ReconcileKeystoneService{}
+
+// ReconcileKeystoneService reconciles a KeystoneService object
+type ReconcileKeystoneService struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a KeystoneService object and makes changes based on the state read
+// and what is in the KeystoneService.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileKeystoneService) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling KeystoneService")
+
+	// Fetch the KeystoneService instance
+	instance := &keystonev1.KeystoneService{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: instance.Spec.AuthURL,
+		Username:         instance.Spec.Username,
+		Password:         instance.Spec.Password,
+		TenantName:       instance.Spec.Project,
+		DomainName:       instance.Spec.DomainName,
+	}
+
+	provider, err := openstack.AuthenticatedClient(opts)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	endpointOpts := gophercloud.EndpointOpts{Type: "identity", Region: instance.Spec.Region}
+	identityClient, err := openstack.NewIdentityV3(provider, endpointOpts)
+
+	// Create new service if ServiceID is not already set
+	if instance.Status.ServiceID == "" {
+		createOpts := services.CreateOpts{
+			Type:    instance.Spec.ServiceType,
+			Enabled: &instance.Spec.Enabled,
+			Extra: map[string]interface{}{
+				"name":        instance.Spec.ServiceName,
+				"description": instance.Spec.ServiceDescription,
+			},
+		}
+
+		service, err := services.Create(identityClient, createOpts).Extract()
+		if err != nil {
+			reqLogger.Error(err, "error")
+			return reconcile.Result{}, err
+		}
+
+		// Set ServiceID in the status
+		reqLogger.Info("instance.Status.ServiceID", "ServiceID", instance.Status.ServiceID)
+		reqLogger.Info("service.ID", "service.ID", service.ID)
+		if instance.Status.ServiceID != service.ID {
```

Comment:
isn't instance.Status.ServiceID is always "" under this conditional path?

Reply:
still unclear to me what updates instance.Status.ServiceID to become not an empty string...